Let mutter wayland recover from vt-switches when using the EGLStream backend

mvicomoya opened the pull request:
These changes will make mutter wayland recover from vt-switches when using the EGLStream backend.

Also, a few minor enhancements are included.

Note these will work with NVIDIA's r375.39 and r378.13 drivers (and future versions).

owtaylor committed:
CoglGPUInfo - fix check for NVIDIA

NVIDIA drivers have a vendor of "NVIDIA Corporation" not "NVIDIA".
Check for both in case older drivers did use "NVIDIA"

https://bugzilla.gnome.org/show_bug.cgi?id=779039

owtaylor committed:
CoglWinsysGLX: factor out some duplicated code

Add a helper function for repeated calls to clock_gettime(CLOCK_MONOTONIC)

https://bugzilla.gnome.org/show_bug.cgi?id=779039

owtaylor committed:
Usability of SGI_video_sync is per-display not per-renderer

As previously commented in the code, SGI_video_sync is per-display, rather
than per-renderer. The is_direct flag for the renderer was tested before
it was initialized (per-display) and that resulted in SGI_video_sync
never being used.

https://bugzilla.gnome.org/show_bug.cgi?id=779039

owtaylor committed:
Fix the get_clock_time() without GLX_OML_sync_control

When we don't have GLX_OML_sync_control, we still can set the
frame presentation time, but we always use the system monotonic time,
so return that from get_clock_time().

https://bugzilla.gnome.org/show_bug.cgi?id=779039

owtaylor committed:
For NVIDIA proprietary drivers, implement sync events with a thread

It's a good guess that the buffer swap will occur at the next vblank,
so use glXWaitVideoSync in a separate thread to deliver a sync event
rather than just letting the client block when frame drawing, which
can signficantly change app logic as compared to the INTEL_swap_event
case.

https://bugzilla.gnome.org/show_bug.cgi?id=779039

owtaylor committed:
Add cogl_xlib_renderer_set_threaded_swap_wait_enabled()

Because the threaded-swap-wait functionality requires XInitThreads(),
and because it isn't clear that it is a win for all applications,
add a API function to conditionally enable it.

Fix the cogl-crate example not to just have a hard-coded dependency
on libX11.

https://bugzilla.gnome.org/show_bug.cgi?id=779039

piotrdrag committed:
Use Unicode in translatable strings

See https://developer.gnome.org/hig/stable/typography.html

https://bugzilla.gnome.org/show_bug.cgi?id=772218

nwnk committed:
Use eglGetPlatformDisplay

Different libEGL will do different things for eglGetDisplay since it has
to guess what kind of display it's been handed. Better to just use the
API that makes it explicit.

Signed-off-by: Adam Jackson <ajax@redhat.com>

https://bugzilla.gnome.org/show_bug.cgi?id=772422

rtcm committed:
wayland-xdg-shell: Handle the wl_output on the set_fullscreen request

This makes us fullscreen wayland windows on the requested monitor.

https://bugzilla.gnome.org/show_bug.cgi?id=772525

garnacho committed:
clutter/evdev: Add ClutterInputDeviceEvdev::device-matrix property

And transform absolute events using this matrix. This property is
akin to the "Coordinate Transformation Matrix" property in X11,
and will be used to map tablets/touchscreens to outputs, favoured
over the libinput matrix which is meant for calibration.

https://bugzilla.gnome.org/show_bug.cgi?id=774115

garnacho committed:
clutter/evdev: Add ClutterInputDeviceEvdev::output-aspect-ratio

When != 0, this property will express the W:H ratio of the desired
output area (be it one monitor or the span of all monitors). At
the time of translating coordinates, coordinates will be skewed so
that the input area of the tablet is a rectangle of the given ratio.

Events happening in the input areas that fall outside the output
aspect ratio will be clamped to the nearest coordinate in the rect.

If the ratio is 0, the whole input area of the tablet will be used
and no coordinate skewing will happen.

https://bugzilla.gnome.org/show_bug.cgi?id=774115

garnacho committed:
backends/native: Use ClutterInputDeviceEvdev::device-matrix for map-to-output

The libinput calibration matrix should only be used for calibration purposes,
this ClutterInputDeviceEvdev property has been added in place.

https://bugzilla.gnome.org/show_bug.cgi?id=774115

garnacho committed:
backends/native: Use libinput calibration matrix for calibration

We couldn't properly merge output-mapping matrix and calibration into
one. Now that libinput calibration matrix is free to use, we can
actually implement tablet calibration with it.

https://bugzilla.gnome.org/show_bug.cgi?id=774115

garnacho committed:
backends/native: Implement keep-aspect

Using ClutterInputDeviceEvdev::output-aspect-ratio. This only applies
to devices which are not calibratable, so again we need to implement
this at the toolkit level.

https://bugzilla.gnome.org/show_bug.cgi?id=774115

rtcm committed:
constraints: Make zero sized windows be placed in the correct monitor

Wayland windows can be zero sized until clients attach a buffer, but
our rectangle code doesn't deal with this case well, in particular,
meta_screen_get_monitor_for_rect() might end up choosing the wrong
monitor for a zero sized rectangle since
meta_rectangle_contains_rect() considers a zero sized rectangle at the
right or bottom edges of another rectangle (the monitor's) to be
contained there.

Since out size limits constraint will enforce a minimum size of 1x1,
we might as well enforce that when setting up the constraint info so
that the correct monitor gets chosen and the single monitor constraint
doesn't move these windows to the wrong one.

https://bugzilla.gnome.org/show_bug.cgi?id=772525

rtcm committed:
window: Make move_to_monitor work for zero sized and offscreen windows

Wayland windows are initially zero sized until clients commit the
first buffer. Despite being invisible, clients are allowed to request
such windows to be fullscreened on a specific output before they
attach the first buffer which means we need to be able to move them.

meta_window_move_to_monitor() doesn't handle this case because these
windows' initial monitor is a placeholder since their initial
coordinates are 0,0+0+0, which results in us using a rectangle as
old_area for meta_window_move_between_rects() that might be to the "right"
of the window causing the move to go further out of the visible
screen's coordinates. This is later "corrected" by the constraints
system but the window might end up in the wrong monitor.

To fix this, we can make meta_window_move_between_rects() accept a
NULL old_area, meaning that we move the window to the new_area without
trying to keep a relative position.

https://bugzilla.gnome.org/show_bug.cgi?id=772525

jadahl committed:
MetaMonitorManagerKms: Don't leak tile property blob

Also warn when we don't understand the property blob we got.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

rtcm committed:
clutter/evdev: Avoid losing key repeat timestamp resolution

Commit 9214d5029c630e6ed8fd9793f6bcb0a0ae451576 changed the notify*
API to use microseconds and we already have a microsecond time source
here so we can use the timestamp directly without losing resolution at
this layer.

https://bugzilla.gnome.org/show_bug.cgi?id=774989

jadahl committed:
Rename MetaMonitorInfo MetaLogicalMonitor

In preparation for further refactorizations, rename the MetaMonitorInfo
struct to MetaLogicalMonitor. Eventually, part of MetaLogicalMonitor
will be split into a MetaMonitor type.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
screen: Remove unused function

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
screen: Fly-by coding style fixes

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
wayland: Make surface role vfunc better named

It checks whether a surface is on a given "logical monitor", not
output. Output here is the Wayland name for the same thing, but should
not be confused with MetaOutput.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

rtcm committed:
clutter/evdev: Dispatch libinput before generating key repeat events

Since both the libinput event source and the key repeat timer have the
same priority, the order in which both handlers are called is
arbitrary if both sources are ready on the same poll return. This
means that sometimes we generate key repeats when there's already a
real key event queued on libinput that would cancel the repeat timer
if only it was processed before.

One solution would be lowering the repeat timer source priority a
notch lower than the libinput source but that would mean that a steady
stream of events from libinput (e.g. pointer device motion) would
prevent any key repeats to happen.

Instead, we can fix this problem by trying to dispatch libinput from
the key repeat timer and checking if the timer source has been
destroyed before generating more key repeats.

https://bugzilla.gnome.org/show_bug.cgi?id=774989

jadahl committed:
Make libmutter and friends parallel installable

In order to minimize the amount of breakage, while at the same time
making it easier to make backward incompatible changes needed to
continue turning libmutter into a capable Wayland compositor, make the
libmutter and friends (libmutter-clutter, libmutter-cogl*) parallel
installable by adding a version number to the name. This changes
various filenames, for example what previously was libmutter.so is now
libmutter-0.so (assuming the version for now is 0), and
libmutter-clutter-1.0.so is now libmutter-clutter-0.so. The pkg-config
filenames and GObject introspection has been renamed to reflect this as
well.

This enables a downstream compositor rely on a specific version of the
libmutter API, while gracefully handling API/ABI changes by having to
update to the new version at their own pace.

https://bugzilla.gnome.org/show_bug.cgi?id=777317

jadahl committed:
MetaScreen: Remove redundant copy of logical monitors array

It was just pointer to the actual list; having to synchronize a list of
logical monitors with the actual monitors managed by the backend is
unnecessary.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaMonitorManager: Track primary monitor by pointer instead of index

Instead of keeping an index in an array to keep track of the current
primary logical monitor, just keep a pointer instead.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaScreen: Remove const qualifiers from monitor getters

They will cause issues when monitors are stared in a GList.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaMonitorManager: Add API turning monitor index into logical monitor

This is the current equivalent of looking up the logical monitor in the
logical monitor array using the number, but eventually that will be
deprecated, and before that done differently, so add a temporary helper
for the places that has not been ported yet.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Add API for getting work area given a logical monitor

To complement the current API which takes an index referencing a
logical monitor in the logical monitor array, add API that takes a
direct reference to the logical monitor itself. The intention is to
replace the usage of the index based API with one that doesn't rely on
internal implementation details.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
core/place: Use logical monitor references instead of indices

Don't rely on the API exposing implementation details and replace it
with the API that takes direct references.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaWindow: Clear fullscreen monitors when monitors change

The fullscreen monitors state is set given a set of xinerama monitor
identification numbers. When the monitor configuration changes (e.g. by
a hotplug event) these are no longer valid, and may point to
uninitialized or unallocated data. Avoid accessing
uninitialized/unallocated memory by clearing the fullscreen monitor
state when the monitor configuration changes.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
window: Keep track of fullscreen monitors directly

Instead of keeping around array indexes, keep track of them by storing
a pointer instead. This also changes from using an array (imitating the
X11 behaviour) to more explicit storing.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
core/events: Don't shadow variable name

Don't use the same variable name for two different things (gesture
tracker vs cursor tracker).

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
core/events: Only get MetaBackend once in meta_display_handle_event

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Move cursor tracker ownership to the backend

Let the backend initialize the cursor tracker, and change all call
sites to get the cursor tracker from the backend instead of from the
screen. It wasn't associated with the screen anyway, so the API was
missleading.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
screen: Remove meta_screen_get_current_monitor_for_pos from public API

It was not used by anyone, and had strange caching behaviour to begin
with.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Put logical monitor grid lookup API in MetaMonitorManager

Don't falsely assume the screen has anything to do with the monitor
grid, and thus move the related API to MetaMonitorManager.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
backends: Use macros for declaring the backend types

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Move more logical monitor getters from the screen to the monitor manager

Turning a rectangle into a logical monitor also has nothing to do with
the screen (MetaScreen) so move it to MetaMonitorManager which has that
information.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
display: Use logical monitor API when requesting pad OSD

Don't use the array index based logical monitor API and switch to the
one directly referencing logical monitor instances.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Move current logical monitor retrieval to the backends

The method used for getting the current logical monitor (the monitor
where the pointer cursor is currently at) depends on the backend type,
so move that logic to the corresponding backends.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Let the MetaWindow calculate its main logical monitor itself

There is no point in going via a MetaScreen API to calculate the "main
monitor" of a window, so let the window do that calculation itself.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Move logical monitor neighbor API into monitor manager

Move the last piece of monitor grid getter API to the monitor manager
away from MetaScreen. The public facing API are still there, but are
thin wrappers around the MetaMonitorManager API.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Make logical monitor from point API take floats

Change meta_monitor_manager_get_logical_monitor_at() to use floats,
replace users of meta_monitor_manager_get_monitor_at_point() to use the
API that returns a logical monitor and remove the now unused function.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaBackend: Move fields from MetaBackend into MetaBackendPrivate

Lets prepare for being able to declare the MetaBackend type using
the macro helper.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
cursor-tracker: Don't cache a private pointer to the cursor renderer

This means we don't rely on contruction order any more.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
screen: Don't use monitor indices in check_fullscreen_func

Rewrite check_fullscreen_func to not use indexes (and
offset-index-as-pointer) tricks. This also removes the usage of an API
constructing temporary logical monitor arrays carrying indices.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
window: Remove public meta_window_get_all_monitors API

It's not used anywhere and used monitor indices which is unoptimal.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Refactor tiled monitor assembly code

Refactor the tiled monitor assembly code (that constructs a logical
monitor out of tiling information. Part of the reason is to move away
from array based storage, part is to make the code easier to follow,
and part is to separate logical monitor construction from list
manipulation.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Store logical monitors in a GList

Instead of storing the logical monitors in an array and having users
either look up them in the array given an index or iterate using
indices, put it in a GList, and use GList iterators when iterating and
alternative API where array indices were previously used.

This allows for more liberty regarding the type of the logical monitor.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Turn MetaLogicalMonitor into a GObject

This will make it easier to move various manually managed data and data
now in MetaLogicalMonitor into GObject QData fields.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
workspace: Don't keep state given logical monitor indices

Don't store logical monitor specific state in an array where the index
from the monitor manager is used as index locally. Instead just use
table associating a logical monitor with a monitor specific state
holder, and store the state in there. This way we don't have the
workspace implementation relying on implementation details of other
units.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
screen: Put xinerama monitor index outside of MetaLogicalMonitor

Put the monitor xinerama index in a separate struct that is attached to
the logical monitor using g_object_set/get_qdata(). Eventually this
should be moved to some "X11 window manager" object, but lets keep it
in MetaScreen until we have such a thing.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
core: Add private API to override compositor configuration

Add private API for overriding the compositor configuration, i.e. the
compositor type (X11 WM or Wayland compositor) and backend type. This
will make it possible to add a special test backend used by src/tests/.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
backends/x11: Make the X11 backend derivable

Make the X11 backend derivable so that we can base a nested test backend
on it.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Use special purpose backend in unit tests

Use a special purpose backend when running the unit tests. This will
eventually allow testing things like configuration and hot plugging.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Add basic monitor configuration tests

The added test checks that two monitors are lied out on a row next to
each other.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
wayland: Add API to override display name

Add API that the tests can use in order to override the Wayland display
name.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Use custom Wayland display name

This gets rid of a warning about the Wayland display name already being
taken.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Make monitor unit tests create initial config

This way we have both the test setup and testing in the same file.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Make monitor test case setup more declarative

In order to more easily create more test cases, make the test case
setup more declarative.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Also make test case verification declarative

Set up the expected result in a declarative way in the same place as
the test case setup is declared. This way we have a completely
declarative way to create test cases.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor: Test two crtcs, one output

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor: Test one out of two outputs being off

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor: Test preferred mode

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Initialize tiled logical monitor scale

The logical monitor scale were not initialized properly when created
for a tiled monitor. Set it to the scale of the main monitor tile.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor-manager: Add API to emulate a hotplug event

Let a test case pass a test setup, which will be used as the new state
emulated by the test backend.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor: Add tiled monitor test

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests/monitor: Add HiDPI test

This doesn't test the actual calculation, since that is done in the
backend, it only tests the configuration.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Rename MetaCRTC(Info) to MetaCrtc(Info)

Change to follow the naming convention used everywhere else. This means
changing CRTC to Crtc even though it is an abbreviation.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Fix private header guard

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
egl: Use correct proc to meta_egl_stream_consumer_aquire_attrib

It should be eglStreamConsumerAcquireAttribNV and not
eglStreamConsumerAcquireAttribEXT as the EXT variant is not defined in
any extension.

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
MetaRendererNative: Remove unnecessary ifdef

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
egl: Add EGL_WL_wayland_eglstream helpers

These are to replace the file descriptor hack currently in use.

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
wayland/buffer: Add 'realization' concept to buffers

Instead of having a way to determine the type of a buffer, add a
realization step that implicitly detects the buffer type. This makes it
possible to both realize (i.e. creating needed objects from the buffer)
and determine the type at the same time, which may be the only possible
way (for example, the only way to know whether a buffer is a EGLStream
is to create the EGLStream from it).

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
wayland/buffer: Remove unnecessary break

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
MetaWaylandEglStream: Use EGL_WL_wayland_eglstream instead of fd hack

Use the proposed EGL_WL_wayland_eglstream EGL extension instead of the
file descriptor hack that was used as a temporary solution.

Note that this results in EGL clients will no longer work if they are
running on a Nvidia driver with a version older than 370.

https://bugzilla.gnome.org/show_bug.cgi?id=773629

jadahl committed:
cogl: Just define the needed GLES2 extension ourself

We need a GLES2 extension macro in cogl-texture-2d-gl.c, but we can't
include GLES2/gl2ext.h because it will conflict with things in
GL/glext.h. We can't rely on cogl including anything for us since it'd
only include GLES2/gl2ext.h if OpenGL support was explicitly disabled.

https://bugzilla.gnome.org/show_bug.cgi?id=774891

jadahl committed:
backends: Rename MetaMonitorMode to MetaCrtcMode

The MetaMonitorMode referred to the mode of a CRTC, and with the future
introduction of a MetaMonitor, theh old name would be confusing.
Instead call it what it is.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

rtcm committed:
wayland: Ensure we don't focus xdg_popups iff they're non-grabbing

Commit 4295fdb892d4474b68e98aa783b725f99c1a2d3a made us skip focusing
all xdg_popups instead of just non-grabbing ones as intended. This
means that when unmanaging a window we might select a xdg_popup window
to focus (in meta_stack_get_default_focus_window() ) but then since we
don't actually focus it we go on unmanaging the focused window which
triggers an assertion, as it should.

To avoid this and still fixing bug 771694 we can make use of the
MetaWindow->input property for non-grabbing xdg_popup windows since
their semantics, in this regard, are the same as no input X11 windows.

This way, when unmanaging a focused window while a xdg_popup is up,
we'll either give focus to the xdg_popup or not select the popup at
all to be focused if it's non-grabbing.

https://bugzilla.gnome.org/show_bug.cgi?id=775986

jadahl committed:
monitor-manager: Move logical monitor into its own file

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Generate a monitor abstraction from outputs and crtcs

Generate a set of "monitors" abstracting the physical concepts. Each
monitor is built up of one or more outputs; multiple outputs being
tiled monitors. Logical monitors will later be built from these.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Check that the monitors are properly generated

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Check that the crtc gets a logical monitor set

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Generate logical monitors from monitors

Instead of using crtcs and outputs to generate logical monitors, use
the ready made monitor abstraction that hides irrelevant things such as
monitor tiling etc.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Add musing about test declarations

https://bugzilla.gnome.org/show_bug.cgi?id=777732

ofourdan committed:
wayland: disconnect mapped signal handler on destroy

Commit 5eb5f72 - wayland: Check surface outputs after mapped state
changes connected the ::mapped signal handler, we need to disconnect it
on destroy to avoid a possible assertion failure in
update_surface_output_state()

https://bugzilla.gnome.org/show_bug.cgi?id=776036

jadahl committed:
monitor-manager: Make tiled Xrandr monitors based on monitors

Don't deal with adding/removing tiled Xrandr monitors in the generic
backend, but leave it to the Xrandr backend. The tiled monitor will
itself notify the backend when such a monitor is added and removed.

Tiled Xrandr monitors are now based no MetaMonitor instead of
MetaLogicalMonitor. This means that mirrored tiled monitors will now be
represented correctly.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check tiled monitor adding/removing notification

Check that the moniter manager backends get the tiled monitor
added/removed notification.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
logical-monitor: Don't duplicate output list

Just use the monitor list and the output lists of each monitor.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
logical-monitor: Don't mirror physical dimensions

Don't try to mirror the physical dimension, since that's a property of
one of the monitors, not of the logical monitor. Callers are changed to
deal with choosing the monitor to represent the logical monitor.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
logical-monitor: Don't mirror refresh rate

Let the callers figure out how to associate a logical monitor with a
refresh rate, instead of exposing an arbitrary one.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
logical-monitor: Remove outdated comment

Remove the comment about tiling things being calculated. It doesn't add
anything valuable, and tiling is hidden behind MetaMonitor.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Add monitor modes abstraction

Add "monitor modes" abstracting the modes set on a monitor. On normal
monitors, this directly maps to the CRTC modes, but on tiled monitors,
a monitor mode can consist modes per tiled output.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check monitor modes

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config: Modernize type declaration

https://bugzilla.gnome.org/show_bug.cgi?id=777732

ofourdan committed:
wayland: Preserve the event mask on the root window

A window manager must select the SubstructureRedirect mask on the root
window to receive the MapRequest from the X11 clients and manage the
windows. Without this event mask set, a window manager won't be able to
map any new window.

The Wayland selection code in mutter can change/clear the event mask on
the requestor window from a XSelectionRequest event when the window is
not managed by mutter/gnome-shell.

A rogue or simply buggy X11 client may send a XConvertSelection() on the
root window and mutter will happily change/clear its own event mask on
the root window, effectively turning itself into a regular X11 client
unable to map any new X11 window from the other X11 clients.

To avoid this, simply check that the requestor window is not the root
window prior to change/clear the event mask on that window.

https://bugzilla.gnome.org/show_bug.cgi?id=776128

fmuellner committed:
Bump version to 3.23.3

Update NEWS.

jadahl committed:
MetaMonitor: Add mode spec concept

Add a "mode spec" concept, meaning to be used as a identifier for an
actual monitor mode. It consists of details making a mode unique, i.e.
the total resolution and refresh rate. This will later be used to get
the actual monitor mode (set of one or more CRTC modes).

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Determine preferred monitor mode

Determine the preferred monitor mode given what the preferred CRTC mode
is, and expose it via the API.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Add monitor spec object

A monitor spec object is meant to be used to identify a certain monitor
on a certain output. The spec is unique per actual monitor and connector,
meaning that a monitor that changes from one connector from another
(e.g. HDMI1 to HDMI2) will not be identified as the same. It is meant
to associate for example a configuration entry with an actual monitor.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaMonitor: Add _is_laptop_panel() helper

Add a helper to determine whether the monitor is the laptop panel or
not. To be used by the configuration system.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Use MetaMonitor to check laptop panel status

Instead of looking at the current configuration, just find the
MetaMonitor of the laptop display panel and check the currest status
directly.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Move configuration paths behind 'legacy' functions

In preparation of replacing the configuration system with one working
with high level monitors instead of low level outputs etc, move
configuarion handling code into obviously named function (containing
the word 'legacy'.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Rename read_current_config to read_current_state

The function meta_monitor_manager_read_current_config() was renamed to
meta_monitor_manager_read_current_state() as it does not read any
configuration, but reads the current state as described by the backend.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Move meta_monitor_manager_ function to right place

Move a meta_monitor_manager_ function to meta-monitor-manager.c where
it belongs.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Test suggested-x/y configuration

Test the suggested-x/y output state often set by virtual machines.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Allow monitor mode foreach crtc iterator failable

Allow the function used to iterate over the CRTC modes of a monitor
mode to fail, causing the iteration to fail as well.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Some trivial style fixes

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Add API to find primary monitor

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Add API to get monitor from monitor spec

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Add API to get monitor mode from spec

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager-kms: Handle EDID blob not being ready immediately

Sometimes we hit a race on hot-plug where we try to read the KMS
resources and the EDID blob is not yet ready. This would normally
result in a ENOENT when retrieving the blob. Handle this by retrying
after 50 milliseconds after a hot-plug event. Do this up to 10 times,
and after that give up trying to get the EDID blob and continue with
best effort.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager-kms: Some harmless style fixes

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
wayland/output: Pass 'unknown' subpixel order when mismatch

When a logical monitor constains monitors with different subpixel
ordering, make the wl_output have the subpixel order 'unknown' so that
clients don't make assumptions given only a subset of the monitors of
the given region.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Add getter for vendor and serial

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Track current monitor mode

Track what monitor mode is the current. This is derived from the mode
of the corresponding CRTC's.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
wayland/output: Operate on monitors instead of outputs

Operate on MetaMonitor's instead of MetaOutput's, as the latter may be
only a subset of an actual "monitor" when referring to the physical
computer equipment.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

garnacho committed:
clutter: Add current group mode on pad events

This info can be obtained here right away, so we don't need
backend-dependent paths in the upper layers.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
clutter: Add functions to find out mode switch buttons, and their group

This is done in the upper layers through backend-dependent ways, seems
better to let ClutterInputDevice provide this information.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
clutter: Add function to find out the number of modes for a pad group

This is obtained in backend-dependent ways in the upper layers, seems
better to let ClutterInputDevice provide this information.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
backends: Have meta_input_settings_handle_pad_button() take an event

As all the relevant backends are expected to provide
ClutterPadButtonEvents, it makes no sense to split the information,
plus all other event fields are now available and might be needed
in the future.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
wayland: Clean up MetaWaylandTabletPadGroup

Using the clutter counterparts, some backend-specific code can be removed
from here.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
core: Add MetaDisplay::show-osd signal

And add specific private methods to notify about tablet mapping and mode
switches. The signal allows the mutter side to trigger OSDs in a generic
way.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
backends: Add action label to mode switch buttons

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
backends: Notify tablet mapping changes in the UI

This takes over the older code in g-s-d.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
wayland: Notify tablet mode switches

This will show a fancy OSD so the change is immediately visible.

https://bugzilla.gnome.org/show_bug.cgi?id=771098

garnacho committed:
core: Add MetaDisplay::pad-mode-switch signal

This signal provides the necessary information to let gnome-shell trigger
updates of pad leds/oleds whenever a pad group switches mode, and the
actions associated to buttons do too.

https://bugzilla.gnome.org/show_bug.cgi?id=776543

jadahl committed:
wayland/xdg-shell: Scale window menu coordinates

When the monitor the surface is on has a scale other than 1, the
coordinate of the window menu popup position needs to be scaled, as it
is reported in logical pixels, while the stage is still in physical
pixels.

https://bugzilla.gnome.org/show_bug.cgi?id=776055

Commit:
Updated Norwegian bokmål translation.

jadahl committed:
tests/monitor-unit-tests: Check current monitor mode

Check both that current mode is up to date, and that deriving it again
won't change it.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-manager: Let backend implementations handle inital config

How the initial configuration is done depends on the backend, so let
them do it themself.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Introduce new monitor configuration system

The new monitor configuration system (MetaMonitorConfigManager) aims to
replace the current MetaMonitorConfig. The main difference between the
two is that MetaMonitorConfigManager works with higher level input
(MetaMonitor, MetaMonitorMode) instead of directly looking at the CRTC
and connector state. It still produces CRTC and connector configuration
later applied by the respective backends.

Other difference the new system aims to introduce is that the
configuration system doesn't manipulate the monitor manager state; that
responsibility is left for the monitor manager to handle (it only
manages configuration and creates CRTC/connector assignments, it
doesn't apply anything).

The new configuration system allows backends to not rely on deriving the
current configuration from the CRTC/connector state, as this may no longer be
possible (i.e. when using KMS and multiple framebuffers).

The MetaMonitorConfigManager system is so far disabled by default, as
it does not yet have all the features of the old system, but eventually
it will replace MetaMonitorConfig which will at that point be removed.
This will make it possible to remove old hacks introduced due to
limitations in the old system.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor: Add API to query suggested position

Adds an API to get the position suggested by the backend. This
translates to position advertised by some VM:s, used to hint at a
position making the position more natural (i.e. placed similarly to how
it may be placed on the host desktop).

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config-manager: Add support for suggested configuration

Create a suggested configuration, if such configuration is provided.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Add test for handling too few CRTCs

Check that configuration when there are more outputs than CRTCs does
something sane.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

rtcm committed:
meta-input-settings: Avoid setting key repeat delay or interval to 0

Since doing so causes either errors or misbehavior.

https://bugzilla.gnome.org/show_bug.cgi?id=776919

rtcm committed:
wayland/keyboard: Avoid a division by zero

We don't further sanitize the values since the protocol allows for
everything as long as it's non-negative.

https://bugzilla.gnome.org/show_bug.cgi?id=776919

jadahl committed:
monitor-manager: Move UpClient signal listening into monitor manager

Move the UpClients notify::lid-is-closed signal handling into
MetaMonitorManager, and put the getter behind a vfunc. This means
Placing it behind a vfunc allows custom backends to implement it
differently; for example the test backend can mock the state.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Mock laptop lid state

For now, always assume the laptop lid is open, as all test cases assume
this.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Add laptop lid switch test case

Test that the laptop panel is disabled if the lid is closed.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config-manager: Handle laptop lid being closed

Handle configuring when the laptop lid is closed. This is so far
handled by creating a linear configuration while ignoring the laptop
panel. Changing the current configuration will come later.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Test starting with lid closed

Add a test case that checks that configuration works when the lid is
initialy closed then later opened. This test case is disabled when the
legacy configuration is used as it does not handle that situation.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
MetaCursorRendererNative: Always force set hw cursor the first time

The initial state of the hardware cursor is not known, so always force
update it the first time we update the cursor. Do this by changing the
'force' flag of update_hw_cursor() to an 'invalidated' hw cursor state.

https://bugzilla.gnome.org/show_bug.cgi?id=771056

jadahl committed:
monitor-config-manager: Add MetaMonitorsConfig creation helper

Add a meta_monitors_config_new() helper. It's exposed outside of
meta-monitor-config-manager.c already, as it'll be used externally in a
later commit.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Use g_assert_cmp* macros when suitable

Use the g_assert_cmp(int|uint|...) macros when comparing integers and
unsigned integers. This means that the mismatched numbers are printed
in the test report.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Add lid-closed with no external monitor test case

Test that when the laptop lid is closed, but no external monitors are
connected, that we still have a mode set on the laptop panel.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config-manager: Handle headless setup gracefully

Handle headless setup gracefully by having no logical monitors. This
commit only makes the monitor management code deal with it; other areas
may still not be able to handle it.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
backend: Only center pointer when not headless

When we are headless, there is no logical monitor to try to center the
pointer on, so just skip it.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Test headless mode

This tests only the monitor configuration and basic functionality. It
does not test anything related to window management and Wayland client
interaction.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
unit-tests: Pass argc/argv directly to g_test_init

Don't let a dummy option context consume the arguments; just let the
GLib test suite do it. It'll handle the basic command line arguments
and allow doing things such as specifying what test to run.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
Introduce MetaMonitorConfigStore

MetaMonitorConfigStore provides an XML storage mechanism for
MetaMonitorConfigManager. It stores configuration files defined in the
same level as the MetaMonitorsConfig format, i.e. refers to high level
"monitors" and "monitor modes" instead of connectors and CRTCs.

Only reading custom files are implemented and so far unused.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Add MetaMonitorConfigStore tests

Add tests that check that a XML configuration file results in correct
MetaMonitorsConfig objects.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-store-unit-tests: Add test for vertical placement

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config-manager: Prefer to use stored config

If not explicitly set by the backend, prefer to use the stored config
instead of creating a new one, if available.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Move monitor test helpers to a helper file

Both the monitor unit tests and monitor store unit tests will want to
check whether the config manager is used and set custom configuration
files.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Make hotplug_mode_update optional

Add a flag to the test setup creation function enabling the caller to
specify whether a stored config should be used. This is done by
changing the value of the hotplug_mode_update MetaOutput field,
normally used by VMs to do the same.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check that a stored config is preferred

Check that the stored config (where the logical monitors are placed
vertically) is preferred to the otherwise generated linear config.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-store-unit-tests: Check the primary and presentation flags

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check handling of the primary presentation flags

The main output of the main (first) monitor of a primary logical
monitor should have the MetaOutput::is_primary field set; all other
outputs should have it not set.

All outputs associated with a logical monitor with presentation set
should have MetaOutput::is_presentation set. No other outputs should
have it set.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Add monitor and monitor store test for custom primary

Test that having a custom primary logical monitor is configured as
expected.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
legacy-monitor-config: Fix primary output state

The MetaOutput::is_primary state was not correctly managed in two cases:

 * for tiled monitors, the primary state got overridden when setting
   the preferred resolution
 * for laptop lid, it was not set if the laptop panel happened to be
   the first output

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-config-manager: Add underscanning to MetaMonitorConfig

Set the underscanning state via the monitor config struct so that it
can be configured.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
legacy-monitor-config: Init is_underscanning config field from output

Don't loose the is_underscanning field when creating a new configuration.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-store: Add support for monitor underscanning

Add a <monitor> subelement <underscanning> for configuring
output underscanning.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
tests: Add monitor store and config test for output underscanning

https://bugzilla.gnome.org/show_bug.cgi?id=777732

hadess committed:
egl: Fix GCC warning

src/backends/meta-egl.c: In function ‘set_egl_error’:
src/backends/meta-egl.c:144:16: error: format not a string literal and no format arguments [-Werror=format-security]
      error_str);
      ^~~~~~~~~

https://bugzilla.gnome.org/show_bug.cgi?id=777389

jadahl committed:
monitor-unit-tests: Check monitor mode CRTC mode positions

Check that the crtc mode has the correct intra-monitor position. In
effect, this tests that the CRTCs in a tiled monitor are configured with
the correct mode on the correct position.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check that each CRTC mode is correct

This is partly redundant given that the current monitor mode is checked
against the CRTC mode, but this also checks the disabled CRTCs. Later
the configured mode position and transform will be checked.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Move logical monitor checking into separate function

Split the big check function, makes things easier to read.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

jadahl committed:
monitor-unit-tests: Check monitor <-> logical monitor relationships

Make sure that each logical monitor owns the expected actual monitors.
This currently needs special care when dealing with laptop lid the
configuration, as the MetaMonitorConfigManager path still deosn't
handle restoring the previous configuration, meaning the logical
monitor with the external monitor will continue being primary.

https://bugzilla.gnome.org/show_bug.cgi?id=777732

garnacho committed:
compositor: Avoid thaw on inconsistent effect_completed calls

If the meta_window_actor_effect_completed() triggers inconsistent
accounting, there's also high chances that the thaw call will be
unexpected at this time too, which will lead to a g_error().

This makes mutter more lenient to effect_completed() calls of the
right type (i.e. those triggering freeze/thaw) being performed more
times than necessary in the upper parts. A warning will be issued,
but the process won't abort.

https://bugzilla.gnome.org/show_bug.cgi?id=777691

jadahl committed:
backends/x11: Split up X11 backend into Cm and Nested

Split up the X11 backend into two parts, one for running as a
Compositing Manager, and one for running as a nested Wayland
compositor.

This commit also cleans up the compositor configuration calculation,
attempting to make it more approachable.

https://bugzilla.gnome.org/show_bug.cgi?id=777800

piotrdrag committed:
Fix a translator comment

They need to be exactly one line above a string to show up in .po files.

jadahl committed:
backends/x11: Only apply keymap when not running nested

https://bugzilla.gnome.org/show_bug.cgi?id=777800

jurf committed:
Update shadows to match Adwaita better

This should make Mutter shadows almost indifferentiable to the ones made
by Adwaita. Based on an initial patch from Maxim <dead.555@bk.ru>.

https://bugzilla.gnome.org/show_bug.cgi?id=744667

fmuellner committed:
window-actor: Special-case COMBO window's shadow class

We currently don't have any shadow class for combo box popups,
which means the default shadow of normal windows is used. That's
clearly odd given that the two are very different, and isn't
consistent with GTK+-3's client-side shadows for popups. While
we could add a dedicated shadow class, the designers are fine
with reusing the existing shadow for dropdown-menus, so let's
do that.

https://bugzilla.gnome.org/show_bug.cgi?id=744667

Commit:
egl: Add stream-related error descriptions

This change adds descriptions for the following errors to
get_egl_error_str():

  - EGL_BAD_STREAM_KHR
  - EGL_BAD_STATE_KHR
  - EGL_BAD_DEVICE_EXT
  - EGL_BAD_OUTPUT_LAYER_EXT

Commit:
egl: Add EGL_RESOURCE_BUSY_EXT error

This change adds both the error code and description for
EGL_RESOURCE_BUSY_EXT, recently added to the EGL_EXT_stream_acquire_mode
extension:

  https://github.com/aritger/eglstreams-kms-example/blob/master/proposed-extensions/EGL_EXT_stream_acquire_mode.txt

Commit:
MetaRendererNative: Properly handle EGLOutput acquire errors

Whenever an EGLOutput consumer is temporary unable to handle
eglStreamConsumerAcquire() operations (e.g. during a VT-switch),
an EGL_RESOURCE_BUSY_EXT error is generated.

This change adds the appropriate error handling to flip_egl_stream() in
order to recover from such errors.

garnacho committed:
backends: Pass seat ID to evdev ClutterDeviceManager

Clutter assumed seat0 which is most usually, but not always correct.
Add an evdev-backend specific function to allow passing the seat
that will be used for ClutterDeviceManager construction, which we
already obtain in MetaLauncher.

https://bugzilla.gnome.org/show_bug.cgi?id=778092

wang-mengying committed:
Update zh_CN translation

Commit:
keybindings: fix erratic raise_or_lower behavior

Function "handle_raise_or_lower (src/core/keybindings.c)" is called
when running 'raise-or-lower' on a window. This function iterates
through all the windows in the stack to determine if our window is
already on top or obscured. The problem is that the window stack
includes windows in another workspaces and also windows that are
minimized.

https://bugzilla.gnome.org/show_bug.cgi?id=705200

whot committed:
wayland: fix copy/pasto sending tool removed on rings/strips

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

https://bugzilla.gnome.org/show_bug.cgi?id=778262

piotrdrag committed:
Update POTFILES.skip

garnacho committed:
backends: Allow opening /sys/ files on MetaLauncher

libinput may want to access those for fetching LED status, as those are
requested readonly, just forward the request to plain open().

https://bugzilla.gnome.org/show_bug.cgi?id=778472

garnacho committed:
wayland: Keep pointer to cursor sprite on MetaWaylandTabletTool

https://bugzilla.gnome.org/show_bug.cgi?id=778474

garnacho committed:
wayland: Update tool cursor scale when crossing monitors

This makes tool cursors properly scaled on hidpi.

https://bugzilla.gnome.org/show_bug.cgi?id=778474

Marv-CZ committed:
Updated Czech translation

haaninjo committed:
Update Swedish translation

prescott66 committed:
Update Slovak translation

piotrdrag committed:
Update Polish translation

rtcm committed:
MetaInputSettings: allow edge scrolling without 2fg capable devices

We should only force edge scrolling off if two finger is enabled *and*
we actually have two finger capable devices.

https://bugzilla.gnome.org/show_bug.cgi?id=778554

Commit:
Updated Norwegian bokmål translation.

Commit:
Updated Slovenian translation

llandwerlin-intel committed:
cogl: fix get-value-func annotation

Found while trying to enable Cogl 2.0 APIs access from JS.

https://bugzilla.gnome.org/show_bug.cgi?id=778684

jbicha committed:
build: Update src/Makefile.am

https://bugzilla.gnome.org/show_bug.cgi?id=778699

fmuellner committed:
build: Clean up .pc files

fmuellner committed:
Bump version to 3.23.90

Update NEWS.

jadahl committed:
egl: Pass EGL errors via custom error domain

Introduce META_EGL_ERROR and have the error code be EGL error codes.

jadahl commented on src/backends/meta-egl.c:
```diff
@@ -656,23 +656,24 @@ meta_egl_stream_consumer_output (MetaEgl          *egl,
 }
 
-gboolean
+EGLint
```
I would prefer we keep the API style consistent, and use the GError to pass the error. I made a patch that makes the GError carry the EGL error here: https://github.com/jadahl/mutter/commit/0a9a9b6fca32dc636f74a7bdb51e52945992d437
Then on the call site, you'd do if (error->domain == META_EGL_ERROR && error->code == EGL_ERROR_IN_QUESTION) ...

mvicomoya replied:
Sure thing. Will you rebase wip/egldevice on top of wip/meta-egl-error? Or do you want me to just cherry-pick that commit?

jadahl replied:
wip/egldevice is already on master upstream, so no need to rebase.

mvicomoya replied:
Oh, I missed that. Thank you!

jadahl commented on src/backends/native/meta-renderer-native.c:
```diff
@@ -532,32 +533,46 @@ flip_egl_stream (MetaRendererNative *renderer_native,
+      if (egl_error == EGL_RESOURCE_BUSY_EXT)
+        {
+          g_warning ("Failed to flip EGL stream (%s)", error->message);
+          *fb_in_use = TRUE;
```
This should probably not set to TRUE, since the call didn't have any effect. Then again, "fb_in_use" is not a thing in EGLDevice paths (as its ignored) but it seems I had it set always anyway. I also think the warning is not needed. It's expected to happen, so it shouldn't go into logs as warnings.

jadahl commented on src/backends/native/meta-renderer-native.c:
```diff
@@ -597,9 +612,11 @@ meta_onscreen_native_flip_crtc (MetaOnscreenNative *onscreen_native,
+                           onscreen_native,
+                           flip_closure,
+                           fb_in_use))
+          onscreen_native->pending_flips++;
```
nit: wrong indentation

garnacho committed:
autogen: Drop usage of gnome-common

dooteo committed:
Update Basque language

dmustieles committed:
Update Spanish translation

fmuellner committed:
display: Check we have a screen before freeing it

The reason for the display to be closed may be meta_screen_new()
returning NULL, in which case we don't have a screen to free.
Avoid a segfault on exit by adding a proper check.

https://bugzilla.gnome.org/show_bug.cgi?id=778831

garnacho committed:
clutter: Force scale=1 in tests/conform

Device pixel sizes are compared in several places, which breaks
some tests on hidpi setups.

claudep committed:
Updated French translation

mariobl committed:
Update German translation

Commit:
Updated Norwegian bokmål translation.

piotrdrag committed:
Update Polish translation

Commit:
Update Indonesian translation

haaninjo committed:
Update Swedish translation

rtcm committed:
x11/xprops: Plug a few memory leaks

Commits 6dbec6f8, 734402e1 and f041b35b introduced memory leaks by
switching to returning copies instead of the original buffers but
forgetting to free those original buffers.

Some error cases were also not freeing the ->prop buffer as they
should.

https://bugzilla.gnome.org/show_bug.cgi?id=642652

MiloCasagrande committed:
Update Italian translation

owtaylor committed:
Call cogl_xlib_renderer_set_threaded_swap_wait_enabled()

Set up things so that if the INTEL_swap_event extension is not present,
but the driver is known to have good thread support, we use an extra
thread and call glXWaitVideoSync() in the thread. This allows idles
to work properly, even when Mutter is constantly redrawing new frames;
otherwise, without INTEL_swap_event, we'll just block in glXSwapBuffers().

https://bugzilla.gnome.org/show_bug.cgi?id=779039

mvicomoya commented:
Addressed comments on top of upstream master and re-sent as a bugzilla bug: https://bugzilla.gnome.org/show_bug.cgi?id=779112
